File: c/problems/problem#2560.c
#include <stdbool.h>
#include <stdlib.h>

int compare(const void *a, const void *b) { return (*(int *)a - *(int *)b); }

int minCapability(int *nums, int n, int k) {
  int *vals = malloc(n * sizeof(int));
  for (int i = 0; i < n; i++) {
    *(vals + i) = *(nums + i);
  }
  qsort(vals, n, sizeof(int), compare);
  int l = 0, r = n - 1;
  while (l < r) {
    int m = (l + r) / 2;
    bool flag = false;
    int count = 0;
    for (int i = 0; i < n; i++) {
      if (flag) {
        flag = false;
        continue;
      }
      if (*(nums + i) <= *(vals + m)) {
        flag = true;
        count++;
      }
    }
    if (count >= k) {
      r = m;
    } else {
      l = m + 1;
    }
  }
  int ans = vals[l];
  free(vals);
  return ans;
}

File: go/problem#2560.go
package main

import "slices"

func minCapability(nums []int, k int) int {
	n := len(nums)
	vals := []int{}
	vals = append(vals, nums...)
	slices.Sort(vals)
	l, r := 0, n-1
	for l < r {
		m := (l + r) / 2
		count := 0
		f := false
		for _, x := range nums {
			if f {
				f = false
				continue
			}
			if x <= vals[m] {
				f = true
				count++
			}
		}
		if count >= k {
			r = m
		} else {
			l = m + 1
		}
	}
	return vals[l]
}
